src/polytopes/segment.jl

```diff
@@ -6,13 +6,23 @@
     Segment(p1, p2)
 
 An oriented line segment with end points `p1`, `p2`.
+The segment can be called as `s(t)` with `t` between
+`0` and `1` to interpolate linearly between its endpoints.
 
 See also [`Line`](@ref).
 """
 struct Segment{Dim,T,V<:AbstractVector{Point{Dim,T}}} <: Polytope{1,Dim,T}
   vertices::V
 end
 
+function (s::Segment)(t)
+  if t < 0 || t > 1
+    throw(DomainError(t, "s(t) is not defined for t outside [0, 1]."))
+  end
```

Comment:
Would this if-else branch create a Union result as we seen in the intersection algorithm? Maybe we can just assume the user will call with any t of interest? Extrapolation should be fine if the performance is compromised with the if-else branch.

Reply:
No, this does not affect performance. The compiler assumes that the branch will never be executed (as far as type inferrence is concerned) because it throws an error. I benchmarked it locally and there is no performance loss. If you call `asin(-2)`, you can see that it is done for trigonometric functions in `Base`, so I think there is no downside to making this check.

Reply:
In the intersection algorithm, it had an impact only when `@error` was present in the branch, because that is not an exception and it actually returned a value. When transformed to an `error` (to throw an exception), the inference worked fine. Though in the end, you just removed the `@error` (which I agree was a good thing), so we did not have the trouble anymore.

Reply:
From `Base` at `special/trig.jl:425`:
```julia
@noinline asin_domain_error(x) = throw(DomainError(x, "asin(x) is not defined for |x|>1."))
function asin(x::T) where T<:Union{Float32, Float64}
    # Since  asin(x) = x + x^3/6 + x^5*3/40 + x^7*15/336 + ...
    # we approximate asin(x) on [0,0.5] by
    #     asin(x) = x + x*x^2*R(x^2)
    # where
    #     R(x^2) is a rational approximation of (asin(x)-x)/x^3
    # and its remez error is bounded by
    #     |(asin(x)-x)/x^3 - R(x^2)| < 2^(-58.75)
    absx = abs(x)
    if absx >= T(1.0) # |x|>= 1
        if absx == T(1.0)
            return flipsign(T(pi)/2, x)
        end
        asin_domain_error(x)
    elseif absx < T(1.0)/2
        # if |x| sufficiently small, |x| is a good approximation
        if absx < ASIN_X_MIN_THRESHOLD(T)
            return x
        end
        return muladd(x, arc_tRt(x*x), x)
    end
    # else 1/2 <= |x| < 1
    t = (T(1.0) - absx)/2
    return asin_kernel(t, x)
end
```

Reply:
Very nice. It is good to know that the compiler can figure out the throw expressions and avoid these Union types as result.